### DISCUSSION TITLE
pfc_n_cst_datetime of_julian and of_gregorian, used to return "removed days" from year 1582.

### DISCUSSION THREAD

**Discussion by ruanito-pombo** [OpenSourcePFCLibraries/2019#27] [Q&A]
Hi, back in the CodePlex days I've pointed out and also provided a very small fix for a very small issue with julian date conversion I experienced with PFC code it was related to the year 1582 on October 4th to 15th. As per this article: (https://en.wikipedia.org/wiki/1582)

Back in the day i've subimitted a fix for this issue, but then my professional life became so full of challenges to overcome, that I've ended up not following up with this. Now I was looking for codeplex and it was gone (lol) I would like to review it, but I don't own a personal PowerBuilder license, so I cant be of much help here, But I do recall the fix I've made. So I'm just wondering, was that fix included in master branch? I know that isn't a issue that will be experienced by all or even many users, but for some apps like historical piece of arts and/or insurance it may affect.

**Comment by bruce-armstrong** [comment]
You don't need PowerBuilder to look at the code, you can see the source code directly here on Github.  The pfc_n_cst_datetime object is likely where the modification would have been made:

https://github.com/OpenSourcePFCLibraries/2019/blob/master/ws_objects/pfcapsrv/pfcapsrv.pbl.src/pfc_n_cst_datetime.sru

I'm not sure what to look for and you likely do, so you can let me know if you find it.

**Reply by ruanito-pombo** [reply]
Hi Bruce, looking at the link you provided I could confirm that the function: "of_julian" and "of_gregorian" has been changed, I found here on my annotations the earlier version of the "of_julian" and "of_gregorian" code with and IF to disconsider those deleted days from it, here it goes:

----------------------------------------------------------------
of_julian(date ad_source)

LONG		ll_d, ll_m, ll_y
LONG		ll_jd, ll_jm, ll_jy
DECIMAL	ld_date

//Check parameters
If IsNull(ad_source) Then
	Return 0
End If

//Check for invalid date
If Not of_IsValid(ad_source) Then
	Return -1
End If

ll_y = Long(String(ad_source,"yyyy"))
ll_m = Long(String(ad_source,"mm"))
ll_d = Long(String(ad_source,"dd"))

// Fix the 0 year from Gregorian Calendar
IF ll_y = 0 THEN
	ll_y = 1
END IF

// Fix the "inexistent" period from 5 through 14 October, 1582 on gregorian calendar
IF (ll_y = 1582) AND (ll_m = 10) AND (ll_d > 4) AND (ll_d < 15) THEN
	ll_d = 15
END IF

IF ll_m > 2 THEN
	ll_jy = ll_y
	ll_jm = ll_m + 1
ELSE
	ll_jy = ll_y - 1
	ll_jm = ll_m + 13
END IF

ld_date = Truncate((365.25 * ll_jy),0) + Truncate((30.6001 * ll_jm),0) + ll_d + 1720995

IF ((ll_d + 31 * (ll_m + 12 * ll_y)) >= (15 + 31 * (10 + 12 * 1582))) THEN
	ld_date = Truncate((ld_date + 2 - Truncate((0.01 * ll_jy),0) + Truncate((0.25 * Truncate((0.01 * ll_jy),0)),0)),0)
END IF

RETURN Truncate(ld_date,5)

----------------------------------------------------------------

as for the other function "of_gregorian" samething, I found an older version (back when the parameter was still a decimal) of it in my annotations with an IF disconsidering those deleted day from the result, here it goes:

----------------------------------------------------------------
of_gregorian(decimal ad_julian)

LONG		ll_day, ll_month, ll_year, ll_h, ll_m, ll_s
DECIMAL	ld_date, ld_time, ld_jd1

//Check parameters
If (IsNull(ad_julian)) OR (ad_julian < 1721424) Then
   datetime ldt_null
   SetNull(ldt_null)
   Return ldt_null
End If

// Correction for Gregorian calendar from 1582 October 5 through 14
IF ld_date = 2299160 THEN
	ld_date = ld_date + 1
END IF

// Extract Date and Time portions
ld_date = Truncate(ad_julian,0)
ld_time = ad_julian - ld_date

IF ld_date >= 2299161 THEN
	ld_jd1 = ld_date + 1 + Truncate((((ld_date - 1867216) - 0.25) / 36524.25),0) - Truncate((0.25 * Truncate((((ld_date - 1867216) - 0.25) / 36524.25),0)),0)
ELSE
	ld_jd1 = ld_date
END IF

ll_day   = Truncate(((ld_jd1 + 1524) - (Truncate(((Truncate((6680.0 + (((ld_jd1 + 1524) - 2439870) - 122.1) / 365.25),0)) * 365.25),0)) - Truncate((Truncate((((ld_jd1 + 1524) - (Truncate(((Truncate((6680.0 + (((ld_jd1 + 1524) - 2439870) - 122.1) / 365.25),0)) * 365.25),0))) / 30.6001),0) * 30.6001),0)),0)
ll_month = Truncate((Truncate((((ld_jd1 + 1524) - (Truncate(((Truncate((6680.0 + (((ld_jd1 + 1524) - 2439870) - 122.1) / 365.25),0)) * 365.25),0))) / 30.6001),0) - 1),0)

IF ll_month > 12 THEN
	ll_month = ll_month - 12
END IF

ll_year = Truncate(((Truncate((6680.0 + (((ld_jd1 + 1524) - 2439870) - 122.1) / 365.25),0)) - 4715),0)

IF ll_month > 2 THEN
	ll_year = ll_year - 1
END IF

// Calculate the Time
ll_h = Truncate((ld_time * 24.0),0)
ll_m = Truncate(((ld_time * 24.0 - ll_h) * 60.0),0)
ll_s = Round((((ld_time * 24.0 - ll_h) * 60.0 - ll_m) * 60.0),0)

RETURN DateTime(Date(ll_year, ll_month, ll_day), Time(ll_h, ll_m, ll_s))
----------------------------------------------------------------

Back in the day, I've ran multiple tests with these codes and I wasn't able to return the excluded days, so I considered a success in my application, I'm now sharing it in hopes that it may help in any way shape or form.


Thanks for your time and efforts in keeping up PowerBuilder great as always.
